pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationDataSource.java

```diff
@@ -82,39 +83,77 @@ default String getHttpHeader(String name) {
 
     /**
      * Check if data from Pulsar protocol are available.
-     * 
+     *
      * @return true if this authentication data contain data from Pulsar protocol
      */
     default boolean hasDataFromCommand() {
         return false;
     }
 
     /**
-     * 
+     *
      * @return authentication data which is stored in a command
      */
     default String getCommandData() {
         return null;
     }
 
+
+    /*
+     * SASL
+     */
+
+    /**
+     *
+     * @return authentication data which will be stored in a command
+     */
+    default byte[] getCommandDataBytes() {
+        return null;
+    }
+
+    /**
+     * set data which will be stored in a command to send to client
+     * used for sasl.
```

Comment:
The interface should not be tied to a specific implementation (eg sasl), but rather abstract the generic needs of multi-step authentication.